### PR TITLE
Watch relevant olivia announcements and attestations

### LINKED ISSUE #70: Monitor for broadcast of non-collaborative settlement transactions
Likely, we want some kind of UI indicator, which transactions of a CFD have been broadcasted and what their state is.

### LINKED ISSUE #110: Create oracle actor that monitors for attestations to price events
We need an actor that interacts with the olivia instance at https://h00.ooo/ and informs other actors about attestations to price events.

1. The CfdActor needs to be told about price attestations so it can expose a choice to the user on settling via the now unlocked CET.
2. The monitoring actor needs to know about the price attestation to start monitoring the correct CET.

The above two can easily be added later because they are just about sending a message to the correct address.

A minimal useful step is to log each price attestation.

### PR DESCRIPTION
**Pull request by luckysori**
Fixes #110.
Fixes #70.

### CODE REVIEW

File: daemon/src/maker_cfd.rs [hunk 1]
```diff
@@ -53,6 +53,7 @@ pub struct Actor {
     current_order_id: Option<OrderId>,
     monitor_actor: Address<monitor::Actor<Actor>>,
     setup_state: SetupState,
+    latest_announcement: Option<oracle::Announcement>,
```
Comment (luckysori) [inline daemon/src/maker_cfd.rs:56]: For now this stores one announcement, but I will change it after #161 is merged.

File: daemon/src/maker.rs [hunk 2]
```diff
@@ -201,6 +202,10 @@ async fn main() -> Result<()> {
                 tokio::spawn(maker_inc_connections_address.attach_stream(listener_stream));
                 tokio::spawn(wallet_sync::new(wallet, wallet_feed_sender));
 
+                oracle::watch_next_announcement(cfd_maker_actor_inbox.clone())
```
Comment (luckysori) [inline daemon/src/maker.rs]: I was unsure if this was the right place to start this task.

File: daemon/src/maker_cfd.rs [hunk 3]
```diff
@@ -311,6 +313,23 @@ impl Actor {
...
 
+        let latest_announcement = self
+            .latest_announcement
+            .to_owned()
+            .context("Unaware of oracle's latest announcement.")?;
+
+        let own_address = ctx
+            .address()
+            .expect("actor to be able to give address to itself");
+        oracle::watch_attestation(
```
Comment (luckysori) [inline daemon/src/maker_cfd.rs]: I was very unsure about starting this task here. Maybe we should wait until after setting up the CFD?

File: daemon/src/maker_cfd.rs [hunk 4]
```diff
@@ -417,6 +436,27 @@ impl Actor {
 
         Ok(())
     }
+
+    async fn handle_oracle_announcements(
+        &mut self,
+        announcements: oracle::Announcements,
+    ) -> Result<()> {
+        tracing::debug!("Updating latest oracle announcements");
+        self.latest_announcement = Some(announcements.0.last().unwrap().clone());
```
Comment (luckysori) [inline daemon/src/maker_cfd.rs:443]: This also changes after #161.

File: daemon/src/maker_cfd.rs [hunk 5]
```diff
@@ -417,6 +436,27 @@ impl Actor {
...
+    }
+
+    async fn handle_oracle_attestation(&mut self, attestation: oracle::Attestation) -> Result<()> {
+        tracing::debug!(
+            "Learnt latest oracle attestation for event: {}",
+            attestation.id
+        );
+
+        todo!(
+            "Update all CFDs which care about this particular attestation, based on the event ID"
```
Comment (luckysori) [inline daemon/src/maker_cfd.rs:455]: Leaving this out of this PR on purpose as discussed this morning.

File: daemon/src/model/cfd.rs [hunk 6]
```diff
@@ -559,6 +559,9 @@ impl Cfd {
                         },
                     }
                 }
+                monitor::Event::CetFinality(_) => {
+                    todo!("Implement state transition")
```
Comment (luckysori) [inline daemon/src/model/cfd.rs:573]: Leaving this out of this PR on purpose as discussed this morning.

File: daemon/src/oracle.rs [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+pub async fn watch_attestation<CFD, M>(
+    announcement: Announcement,
+    cfd_actor_address: xtra::Address<CFD>,
+    monitor_actor_address: xtra::Address<M>,
+) -> Result<()>
+where
+    CFD: xtra::Handler<Attestation>,
+    M: xtra::Handler<Attestation>,
+{
+    let wait_time =
```
Comment (luckysori) [inline daemon/src/oracle.rs]: :mega: Okay, this looks very wrong. 

I would welcome suggestions on how to handle this. Maybe we shouldn't even wait until the `expected_outcome_time` and just try regularly?

File: daemon/src/oracle.rs [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+                "Attestation watcher encountered an unrecoverable error: {:#}",
+                e
+            );
+        }
+    });
+
+    Ok(())
+}
+
+/// Construct the URL of `olivia`'s next `bitmex` event to be
```
Comment (luckysori) [inline daemon/src/oracle.rs]: TODO: Update docstring.

File: daemon/src/oracle.rs [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+        .map(event_url)
+        .collect()
+}
+
+fn next_24_hours(datetime: OffsetDateTime) -> Vec<OffsetDateTime> {
+    let adjusted = datetime.replace_time(Time::from_hms(datetime.hour(), 0, 0).expect("in_range"));
+    (1..=24).map(|i| adjusted + Duration::hours(i)).collect()
+}
+
+/// Construct the URL of `olivia`'s `bitmex` event to be attested
```
Comment (luckysori) [inline daemon/src/oracle.rs]: TODO: Update docstring.

**Comment by thomaseizinger** [comment]
Have a look at what I did here: https://github.com/comit-network/hermes/pull/178

I think triggering a message on an interval is easier to understand and also more flexible because we could in theory trigger the sync manually via another message.

Would you mind transitioning this implementation to something similar?

File: daemon/src/oracle.rs [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+
+/// Construct the URL of `olivia`'s `BitMEX/BXBT` event to be attested
+/// for at the time indicated by the argument `datetime`.
+fn event_url(datetime: OffsetDateTime) -> String {
+    let formatter = format_description::parse(OLIVIA_EVENT_TIME_FORMAT).expect("valid formatter");
+    datetime
+        .format(&formatter)
+        .expect("valid formatter for datetime");
+
+    format!("{}/BitMEX/BXBT/{}/price", OLIVIA_URL, datetime)
```
Comment (thomaseizinger) [inline daemon/src/oracle.rs:181]: Is this correct? I think it is `.price`.

cc @llfourn

File: daemon/src/taker_cfd.rs [hunk 11]
```diff
@@ -166,13 +171,25 @@ impl Actor {
             .send(load_all_cfds(&mut conn).await?)?;
         let cfd = load_cfd_by_order_id(order_id, &mut conn).await?;
 
+        let latest_announcement = self
+            .latest_announcement
+            .to_owned()
+            .context("Unaware of oracle's latest announcement.")?;
+        let nonce_pks = latest_announcement.nonce_pks;
```
Comment (thomaseizinger) [inline daemon/src/taker_cfd.rs:178]: @da-kami You will have to change this to look up the correct announcement based on the ID included in the offer.

File: daemon/src/oracle.rs [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+                    }
+                };
+
+                let attestation = res.json::<Attestation>().await?;
+
+                self.cfd_actor_address
+                    .clone()
+                    .do_send_async(attestation.clone())
+                    .await
+                    .expect("receiving actor to be up");
```
Comment (thomaseizinger) [inline daemon/src/oracle.rs]: I think we should not panic here.

File: daemon/src/oracle.rs [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+                self.monitor_actor_address
+                    .clone()
+                    .do_send_async(attestation)
+                    .await
+                    .expect("receiving actor to be up");
```
Comment (thomaseizinger) [inline daemon/src/oracle.rs]: Same here.

**Review by thomaseizinger — APPROVED** [review]
Thanks, that should do for now :)

**Pull request merged**